radix-engine/src/utils/native_blueprint_call_validator.rs

```diff
@@ -0,0 +1,460 @@
+                $field: ident = $value: expr
+            ),* $(,)?
+        }
+    ) => {
+        struct $ident {
+            $(
+                $field: Option<PackageDefinition>,
+            )*
+        }
+
+        impl $ident {
+            fn new() -> Self {
+                Self {
+                    $(
+                        $field: None,
+                    )*
+                }
+            }
+
+            $(
+                fn $field(&mut self) -> &PackageDefinition {
+                    if let Some(ref schema) = self.$field {
+                        schema
+                    } else {
+                        self.$field = Some($value);
+                        self.$field()
+                    }
+                }
+            )*
+        }
+    };
+}
+
+define_lazy_package_definitions! {
+    struct LazyPackageDefinitions {
+        consensus_manager_package = ConsensusManagerNativePackage::definition(),
+        account_package = AccountNativePackage::definition(),
+        identity_package = IdentityNativePackage::definition(),
+        access_controller_package = AccessControllerNativePackage::definition(),
+        pool_package = PoolNativePackage::definition(),
+        resource_package = ResourceManagerNativePackage::definition(),
+        package_package = PackageNativePackage::definition(),
+        transaction_processor_package = TransactionProcessorNativePackage::definition(),
+        metadata_package = MetadataNativePackage::definition(),
+        royalties_package = RoyaltyNativePackage::definition(),
+        access_rules_package = AccessRulesNativePackage::definition(),
+    }
+}
+
+fn get_blueprint_schema<'p>(
+    package_definition: &'p PackageDefinition,
+    package_address: PackageAddress,
+    blueprint: &str,
+    instruction_index: usize,
+) -> Result<&'p BlueprintSchema, ValidationError> {
+    package_definition.schema.blueprints.get(blueprint).map_or(
+        Err(ValidationError::InvalidBlueprint(
+            instruction_index,
+            package_address,
+            blueprint.to_owned(),
+        )),
+        Ok,
+    )
+}
+
+/// * An `Err` is returned if something is invalid about the arguments given to this method. As an
+/// example: they've specified a method on the account blueprint that does not exist.
+/// * A `None` is returned if the schema for this type is not well know. As an example: arbitrary
+/// packages.
+fn get_arguments_schema<'s>(
+    invocation: Invocation,
+    package_definitions: &'s mut LazyPackageDefinitions,
+    instruction_index: usize,
+) -> Result<Option<(LocalTypeIndex, &'s Schema<ScryptoCustomSchema>)>, ValidationError> {
+    let entity_type =
+        if let Some(entity_type) = invocation.global_address().as_node_id().entity_type() {
+            entity_type
+        } else {
+            return Err(ValidationError::InvalidAddress(
+                instruction_index,
+                invocation.global_address(),
+            ));
+        };
```

Comment:
Minor: just to note that all addresses are entity-type-checked.

Reply:
Gotcha, so you're saying that unwrapping here is safe to do?